applications/finetune-quora-embeddings/benchmark.py

```diff
@@ -0,0 +1,354 @@
+    )
+    sentence_pairs_embedding_2 = embedding_layer(
+        torch.as_tensor(sentence_pairs_elem_2, device="cuda")
+    )
+
+    cosine_scores = util.cos_sim(sentence_pairs_embedding_1, sentence_pairs_embedding_2)
+    predictions = np.diag(cosine_scores.cpu()).tolist()
+    return {name: f(labels, predictions) for name, f in EVALS.items()}
+
+
+@stub.function(
+    image=image,
+    volumes={DATASET_DIR: DATASET_VOLUME},
+    secret=Secret.from_name("cohere"),
+    timeout=86400,
+    gpu=GPU_CONFIG,
+)
+async def benchmark_cohere_roc():
+    from cohere import Client
+    from datasets import load_from_disk
+    from sentence_transformers import util
+    from sklearn.metrics import roc_auc_score
+    import asyncio
+    import time
+    from tqdm import tqdm
+    import numpy as np
+    import torch
+    import torch.nn as nn
+
+    dataset = load_from_disk(f"{DATASET_DIR}/{DATASET_NAME}")
+    test_dataset = dataset["test"]
+    co = Client(os.environ["COHERE_API_KEY"])
+    sem = asyncio.Semaphore(64)
+    num_sentences = 86235  # len(test_dataset)
+    sentence_to_embedding_map = dict()
+    batch_size = 96
+    sentences = get_unique_sentences(
+        test_dataset, sentence_to_embedding_map, batch_size
+    )
+    tqdm_monitoring_bar = tqdm(total=num_sentences)
+
+    async def embed_text(texts, progress_bar: tqdm):
+        async with sem:
+            response = co.embed(
+                texts=texts,
+                model="embed-multilingual-v3.0",
+                input_type="clustering",
+            )
+            progress_bar.update(len(texts))
+            assert len(response.embeddings) == len(texts)
+
+            for item in response.embeddings:
+                assert len(item) == 1024
+            return response.embeddings
+
+    start = time.time()
+    coros = [
+        embed_text(sentence_group, tqdm_monitoring_bar) for sentence_group in sentences
+    ]
+    res = await asyncio.gather(*coros)
+    end = time.time()
+    tqdm_monitoring_bar.close()
+    print(f"Generated embeddings in {end-start}s")
+
+    # Flatten the list of lists and convert to tensor of floats
+    flattened_res = [item for sublist in res for item in sublist]
+    print(f"Processed {len(flattened_res)} embeddings in {end-start}s")
+    embeddings_tensor = torch.tensor(flattened_res).to("cuda")
+    embedding_layer = nn.Embedding.from_pretrained(embeddings_tensor)
+
+    sentence_pairs_elem_1, sentence_pairs_elem_2, labels = extract_unique_sentences(
+        test_dataset, sentence_to_embedding_map
+    )
+    sentence_pairs_embedding_1 = embedding_layer(
+        torch.as_tensor(sentence_pairs_elem_1, device="cuda")
+    )
+    sentence_pairs_embedding_2 = embedding_layer(
+        torch.as_tensor(sentence_pairs_elem_2, device="cuda")
+    )
+
+    cosine_scores = util.cos_sim(sentence_pairs_embedding_1, sentence_pairs_embedding_2)
+    predictions = np.diag(cosine_scores.cpu()).tolist()
```

Comment:
The cosine scores are not the predictions. You should fit a single variable logistic regression to get proper probabilities, Instead of guessing the threshold